tests/rptest/clients/rpk.py

```diff
@@ -1229,6 +1229,8 @@ def _kafka_conn_settings(self):
             "brokers=" + self._redpanda.brokers(),
         ]
         if self._username:
+            # u, p and mechanism must always be all set or all unset
+            assert self._password and self._sasl_mechanism
```

Comment:
Actually I feel like I should just delete this, since a more precise version of this and other invariants are already enforced in the KCS class which is what the property defers to, WDYT? I think I added this before I fully converted rpk.py over to check that this invariant was not being violated before doing the conversion.

Reply:
Ah, yeah. That sounds better 👍

Reply:
@oleiman oh right, now I remember why that check if there: it's needed for typing. 

Typing needs to know that these are non-None since we below append them to a string with `+` (`None` can't be used for that).

We can't move the assert into a function because typing analysis doesn't cross function boundaries.

So to get correct typing you often end up adding these asserts to assert stuff you know for reasons which are opaque to the type checker (or at least you _think_ you know).

The deeper fix for this kind of thing is to express the invariant in the type system, e.g., instead of three separate properties u, p and mechanism where you know that u != None implies p != None (and vice versa), you have a type which has those two properties both non-None and a method which returns this type or None. Then you can't have the mixed None case and don't need the assert to defend against it.

Inside KCS that's how I do it with the SaslCredentials method, but because of the way rpk.py is already written with lots of use of the existing `_username` and `_password` properties it's just easier to throw an assert down here.

Reply:
Makes sense. No worries on my end, mostly just curious and wanting to understand better for future use.

Question: would the type checker do the right thing if you put a return type qualifier on the property def?

Reply:
> Question: would the type checker do the right thing if you put a return type qualifier on the property def?

No, it already infers the right type: it's just that it can't encode a predicate across two distinct variables. E.g., if can know that `u` is `str | None` and `p` is `str | None` but this is actually too wide: its actually that `(u, p)` is `tuple[str, str] | tuple[None, None]`: that the None-ness of one implies the None-ness of the other. With individual u and p properties you can't communicate this to the type checker, you have to do in the caller, e.g., using an assert as above _or_ you need to change your API to not offer separate u and p properties but one method that effective returns say `tuple[str, str] | None`, then after the check `if ret != None` the type you are left with is `tuple[str, str]` which is definitely non-None.

Reply:
ah, yeah, in my head I was thinking that you could effectively make the accessors `str` with assertions but that's obviously not the case. ty for humoring me 😄